File: services/sampleapi/src/SampleApi/Services/NotificationClient.cs
// src/SampleApi/Services/NotificationClient.cs
// HTTP client wrapper for calling the NotificationApi microservice.
// All calls are resilient — failures are logged as warnings but never propagate
// to the caller. This ensures the main API works even if NotificationApi is down.

using System.Text;
using System.Text.Json;

namespace SampleApi.Services;

/// <summary>
/// Typed HttpClient for the NotificationApi microservice.
/// Sends notifications (email, slack, webhook) via POST /api/Notify.
/// URL is configured via NOTIFICATION_API_URL environment variable.
/// </summary>
public class NotificationClient
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };

    private readonly HttpClient _httpClient;
    private readonly ILogger<NotificationClient> _logger;

    public NotificationClient(HttpClient httpClient, ILogger<NotificationClient> logger)
    {
        _httpClient = httpClient;
        _logger = logger;
    }

    /// <summary>
    /// Send a notification to the NotificationApi.
    /// Returns the response from NotificationApi, or null if the call fails.
    /// This method never throws — all exceptions are caught and logged.
    /// </summary>
    /// <param name="type">Notification type: "email", "slack", "webhook"</param>
    /// <param name="message">The notification message body</param>
    /// <param name="recipient">Target recipient (email address, channel, URL)</param>
    public virtual async Task<NotificationResponse?> SendNotificationAsync(
        string type, string message, string recipient)
    {
        try
        {
            var payload = new
            {
                Type = type,
                Message = message,
                Recipient = recipient,
                Timestamp = DateTime.UtcNow
            };

            var json = JsonSerializer.Serialize(payload);
            var content = new StringContent(json, Encoding.UTF8, "application/json");

            _logger.LogDebug(
                "Sending notification: type={Type}, recipient={Recipient}",
                type, recipient);

            var response = await _httpClient.PostAsync("/api/Notify", content);

            if (response.IsSuccessStatusCode)
            {
                var responseBody = await response.Content.ReadAsStringAsync();
                var result = JsonSerializer.Deserialize<NotificationResponse>(
                    responseBody, JsonOptions);

                _logger.LogInformation(
                    "Notification sent successfully: type={Type}, recipient={Recipient}",
                    type, recipient);

                return result;
            }

            // Non-success HTTP status — log but don't throw
            _logger.LogWarning(
                "NotificationApi returned {StatusCode} for type={Type}, recipient={Recipient}",
                response.StatusCode, type, recipient);
            return null;
        }
        catch (HttpRequestException ex)
        {
            // Network error — NotificationApi is unreachable
            _logger.LogWarning(ex,
                "NotificationApi unreachable. Notification not sent: type={Type}, recipient={Recipient}",
                type, recipient);
            return null;
        }
        catch (TaskCanceledException ex)
        {
            // Timeout — NotificationApi took too long to respond
            _logger.LogWarning(ex,
                "NotificationApi timed out. Notification not sent: type={Type}, recipient={Recipient}",
                type, recipient);
            return null;
        }
        catch (Exception ex)
        {
            // Unexpected error — catch-all to prevent caller failure
            _logger.LogWarning(ex,
                "Unexpected error calling NotificationApi: type={Type}, recipient={Recipient}",
                type, recipient);
            return null;
        }
    }
}

/// <summary>
/// Response model from the NotificationApi POST /api/Notify endpoint.
/// </summary>
public class NotificationResponse
{
    /// <summary>Whether the notification was accepted for delivery</summary>
    public bool Success { get; set; }

    /// <summary>Unique ID for tracking the notification</summary>
    public string? NotificationId { get; set; }

    /// <summary>Human-readable status message</summary>
    public string? Message { get; set; }
}

File: services/order-service/src/main/java/com/devsecops/order/service/NotificationClient.java
// order-service/src/main/java/com/devsecops/order/service/NotificationClient.java
// HTTP client for the .NET NotificationApi service.
// Sends notifications (order confirmation, status updates) via the .NET service.
// This is another cross-language call that appears in distributed traces.
// Graceful failure: notification failures don't affect order processing.
package com.devsecops.order.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

import java.util.HashMap;
import java.util.Map;

@Service
public class NotificationClient {

    private static final Logger log = LoggerFactory.getLogger(NotificationClient.class);

    private final RestTemplate restTemplate;
    private final String sampleApiUrl;

    public NotificationClient(
            RestTemplate restTemplate,
            @Value("${services.sampleapi.url}") String sampleApiUrl) {
        this.restTemplate = restTemplate;
        this.sampleApiUrl = sampleApiUrl;
    }

    /**
     * Send a notification via the .NET NotificationApi.
     * Calls POST /api/Notify with type, message, and recipient.
     *
     * @param type      notification type (e.g., "ORDER_CREATED", "ORDER_SHIPPED")
     * @param message   the notification message body
     * @param recipient the notification recipient (email, user ID, etc.)
     */
    public void sendNotification(String type, String message, String recipient) {
        String url = sampleApiUrl + "/api/Notify";
        log.info("Sending notification: type={}, recipient={}, url={}", type, recipient, url);

        try {
            Map<String, String> requestBody = new HashMap<>();
            requestBody.put("type", type);
            requestBody.put("message", message);
            requestBody.put("recipient", recipient);

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            HttpEntity<Map<String, String>> request = new HttpEntity<>(requestBody, headers);

            restTemplate.postForObject(url, request, String.class);
            log.info("Notification sent successfully: type={}, recipient={}", type, recipient);
        } catch (RestClientException e) {
            // Notification failures are non-critical — log and continue
            log.error("Failed to send notification: type={}, recipient={}, error={}",
                    type, recipient, e.getMessage());
        } catch (Exception e) {
            log.error("Unexpected error sending notification: type={}, error={}", type, e.getMessage());
        }
    }
}
